src/components/Tags.tsx

```diff
@@ -20,122 +25,114 @@ const KeyCodes = {
   enter: 13,
 };
 
-type tagstruct = {
-  rank: number;
-  segment: string;
-  tag: { name: string };
-  _id: string;
-};
-
-interface MyProps {
-  tags: { id: string; text: string }[];
-  seg: any;
-  rank: number;
-  forceUpdate: any;
-}
-
-interface MyState {
-  tags: any;
-  suggestions: any;
-  k: number;
-}
-
 const delimiters = [KeyCodes.comma, KeyCodes.enter];
 
-class Tags extends React.Component<MyProps, MyState> {
-  constructor(props: MyProps) {
-    super(props);
-    this.state = {
-      tags: props.tags,
-      suggestions: suggestions,
-      k: 1,
-    };
-
-    this.handleDelete = this.handleDelete.bind(this);
-    this.handleAddition = this.handleAddition.bind(this);
-    this.handleDrag = this.handleDrag.bind(this);
-    this.handleTagClick = this.handleTagClick.bind(this);
-  }
-
-  handleDelete(i: any) {
-    const { tags } = this.state;
-    const delIndex = tags[i].id;
-
-    this.setState({
-      tags: tags.filter((tag: any, index: any) => index !== i),
-    });
-
-    //del from segment
-    for (let j = 0; j < this.props.seg.tags.length; j++) {
-      if (this.props.seg.tags[j].tag._id == delIndex) {
-        this.props.seg.tags.splice(j, 1);
-      }
+const Tags = ({
+  disabled,
+  currentSegment,
+  segmentIndex,
+  rank,
+  updateSegmentAt,
+  refresh,
+}: {
+  disabled: boolean;
+  currentSegment: Segment;
+  segmentIndex: number;
+  rank: number;
+  updateSegmentAt: (index: number, data: Partial<Segment>) => void;
+  refresh: [boolean];
+}): JSX.Element => {
+  const [reactTags, setReactTags] = React.useState([] as ReactTag[]);
+
+  const segmentTagToReactTag = (segTag: SegmentTag): ReactTag => ({
+    text: segTag.tag.name,
+    id: segTag.tag.name,
+  });
+
+  const reactTagToSegmentTag = (reactTag: ReactTag): SegmentTag => ({
+    tag: { name: reactTag.text },
+    rank,
+  });
+
+  const updateTags = (tags: ReactTag[]) => {
+    let currentSegmentTags = currentSegment.tags || [];
+    if (assertModelArrayType<SegmentTag>(currentSegmentTags, 'SegmentTag')) {
```

Comment:
The `assertModelArrayType` method is a helpful utility method that tells typescript that a variable is of a certain type. If the variable is not of that type, an error will be thrown.

If you use intellisense to check the type of `currentSegementTags` outside of the if statement, it will show (`Tag[] | SegmentTag[]`), but if you check the type within the if statement, the `assertModelArrayType` has informed typescript that the variable will definitely be of type `SegmentTag[]`.

You can see the function definition here: https://github.com/cusac/mbt.guide/blob/e81cef8e003e31e0b259652da9ca12457e0213d4/src/types/model.type.ts#L164

And learn more about basic type checking here:
https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

Reply:
Thank you very much. Very helpful.